### ISSUE TITLE
WCAG 2.2 2.5.7 Dragging and its reference to single pointer

### ISSUE THREAD

**Issue by guyhickling** [w3c/wcag#1225]
The wording of the SC is contradictory. It says that path-based gestures (I'll use PBGs for brevity) must have an alternative method using "a single pointer without a path-based gesture". But if you look at the glossary definition of those words "single pointer", it actually includes and therefore appears to explicitly allow PBGs as an alternative. It says:- 

> pointer input that operates with one point of contact with the screen, including ... path-based gestures

This is highly confusing to anyone reading it and trying to work out what it all really means. 

I'm raising this now because the new Dragging SC, SC 2.5.7 in WCAG 2.2, uses the same "single pointer" term, and the glossary definition is referenced in the Understanding. Since 2.5.7 is just like 2.5.1 in that it also seeks to exclude PBGs from the allowed alternative methods (established in an earlier GitHub issue), this glossary definition will now confuse two SCs, not just one.

The only other SC to reference this term is 2.5.2, but that is not about PBGs either. 

Could we delete the words "path-based gestures" from the glossary definition of "single pointer"? None of the SCs want it there and it is just confusing.

(NB: I suggest this should have the 2.1 and 2.2 flags added to it as it affects both.)

**Comment by patrickhlauke** [comment]
the definition for "single pointer" is a bit inaccurate/incorrect - see https://github.com/w3c/wcag/pull/809

however, in this specific case, "single pointer" can include all those interactions (like PBGs), but the SC explicitly says "a single pointer *without a path-based gesture*" so it excludes those. from that point of view, i don't think that's incorrect.

**Comment by guyhickling** [comment]
Unfortunately, 2.5.7 doesn't use that phrase "without a path-based gesture", so PBGs aren't excluded in that SC. In other words 2.5.7 expressly suggests PBGs as an alternative to dragging, which contradicts the intention of 2.5.1 which says they cannot be used without an alternative. Or are you saying that PBGs can be used as the alternative for dragging, but that must then be given another alternative that is not a PGB? That would clearly be unreasonable.

So my point in raising this issue was that none of the three SCs that reference "single pointer" intend to allow PBGs as the alternative, so the single pointer definition should not include them.

**Comment by patrickhlauke** [comment]
or conversely...a single pointer interaction simply means "only uses a single pointer" and does include dragging/swiping, and SCs that don't allow it should say so specifically in the SC text.

**Comment by patrickhlauke** [comment]
or, more specifically: 2.5.7 needs to also explicitly exclude PBG the same way that 2.5.1 had to do. because otherwise, if we redefined it in the definition, wouldn't it be odd to say "single pointer means it's a single mouse, touch or stylus rather than a multitouch gesture...oh, but gestures/drags/swipes with a single pointer are not included in this definition".

**Comment by patrickhlauke** [comment]
I'd retitle the issue here to be something along the lines of "WCAG 2.2 2.5.7 Dragging and its reference to single pointer" or something, and clarify that the normative wording needs to be similar to 2.5.1 and say " single pointer without a path-based gesture"

**Comment by guyhickling** [comment]
I see no point in including something in a definition, then excluding it from every SC that uses it. We have 3 SCs, none of which want to include PBGs, so don't include PBGs in the definition of single pointer. Keep it simple and try not to oppose every suggestion automatically, we are all trying to improve things here. (And the title of this GitHub issue reflects what I am advocating here, so no change needed.)

**Comment by patrickhlauke** [comment]
then the definition itself needs to change not just in content, but in name itself. removing PBG from "single pointer" makes no sense, as a PBG does/can involve a single pointer.

also, unless i'm mistaken, 2.5.2 does/can refer to PBGs as well.

**Comment by guyhickling** [comment]
Yes I agree, calling it 'single pointer' may not have been the best choice. Apart from anything else, a "pointer" is a mouse, pen, finger, headpointer etc. while tabs, clicks and, yes, PBGs, are all actions performed by the pointer.

However it's done and those words are in WCAG 2.1 for SC2.5.1 (and also 2.5.2), so we can't change those SCs. Which is why I was suggesting changing the glossary definition instead to remove the PBGs. 

Since PBGs will always need an alternative that isn't a PBG, I can't envisage a situation where we would actually need to have PBGs in the single pointer definition. It think it would be valid for the WCAG to define any term in a particular way so long as it sticks to that definition throughout, so a single pointer definition that doesn't include PBGs seems to me ok. 

If the definition isn't changed then we would, as you say, have to change SC2.5.7 to add the same words "without a path-based gesture" into the SC itself. That wasn't my favorite solution, simply because it is a bit less simple. But a change is certainly needed here, one way or the other.

**Comment by guyhickling** [comment]
As regards 2.5.2, I have always viewed that as being intended to apply to exactly the same pointer actions as 2.5.1, so not including PBGs. Although it doesn't actually say that, PBGs are not mentioned in the 2.5.2 Understanding doc at all. I'll have to think about that one.

**Comment by alastc** [comment]
Patrick wrote:
> "single pointer" can include all those interactions (like PBGs), but [2.5.2] explicitly says "a single pointer without a path-based gesture" so it excludes those. 

Guy wrote:
> Unfortunately, 2.5.7 doesn't use that phrase "without a path-based gesture", so PBGs aren't excluded in that SC.

For the Dragging SC, it uses it's own definition of 'dragging movements':
https://w3c.github.io/wcag/guidelines/22/#dfn-dragging-movements

The dragging movements includes anything that includes dragging/gestures etc.

For 2.5.1 / 2.5.2, the first uses the definition of single pointer and excludes PBG, the second uses the definition and is not intended to exclude PBG.

I'm not seeing anything to update here?

**Comment by patrickhlauke** [comment]
the core of the problem is that the definition for "single pointer" actually explains what a "single pointer gesture" is (which does include path-based gestures), whereas in the understanding for 2.5.7 https://www.w3.org/WAI/WCAG22/Understanding/dragging.html is says

> The intent of this Success Criterion is to ensure functionality that uses a dragging movement (e.g. sliders, drag-and-drop interfaces) has another single pointer mode of operation without the need for the dexterity required to drag elements.

with "single pointer" linking (though the link is currently broken) to that definition. but it actually means just "using a single pointer input device" (i.e. it doesn't mean "single pointer gestures" in general, including path-based).

my PR here https://github.com/w3c/wcag/pull/809 attempts retrospectively to square this circle, by providing a definition of "single pointer" and then separately expanding what gestures *can* be performed with a single pointer. this would make the current linking from 2.5.7 sort of relevant, as it then refers to the input method itself only, and not to the gestures themselves which do include path-based gestures.

so, I *think* pulling in my proposed PR #809 does the job here as well (though fundamentally yes, then that intent still reads essentially "single pointer, but don't require dragging which is described in the definition of single pointer we just linked you to as one of the possible gestures of single pointer", but that seems fine/clear to me).

**Comment by guyhickling** [comment]
An update is needed to resolve the contradiction in 2.5.7. See my original point at the top of this issue:

> The wording of the SC is contradictory. It says that path-based gestures (I'll use PBGs for brevity) must have an alternative method using "a single pointer without a path-based gesture". But if you look at the glossary definition of those words "single pointer", it actually includes and therefore appears to explicitly allow PBGs as an alternative. It says:-

>>  pointer input that operates with one point of contact with the screen, including ... path-based gestures

> This is highly confusing to anyone reading it and trying to work out what it all really means.

The original solution I suggested needs changing because as Patrick pointed out, 2.5.2 needs the definition of "single pointer" to include PBGs, so my suggestion of changing the definition of single pointer is not an option. Instead, as a better solution, 2.5.7 must exclude PBGs as an alternative. So 2.5.7 needs to say:

> .....can be operated by a single pointer without dragging or path based gestures....

In other words, 2.5.7 needs to add the same exclusion (of PBGs) that 2.5.1 does. 

(Meanwhile, while you're thinking about that, I'll go and look at Patrick's 809 issue that was referenced above.)

**Comment by alastc** [comment]
#809 removes the mention of (single pointer) path-based gestures, so would require an update to 2.5.2 (or an adjustment to the definition).

@guyhickling - "[dragging movements](https://w3c.github.io/wcag/guidelines/22/#dragging-movements)" doesn't mention path based gestures in the SC, so I'm confused. It uses a definition of "dragging movements", which would include any PBGs.

**Comment by patrickhlauke** [comment]
> #809 removes the mention of (single pointer) path-based gestures

to me they were implicit in the mention of "and dragging", but i can add path-based gestures explicitly in that list of examples as well.

**Comment by patrickhlauke** [comment]
> So 2.5.7 needs to say:
> 
> > .....can be operated by a single pointer without dragging or path based gestures....
> 
> In other words, 2.5.7 needs to add the same exclusion (of PBGs) that 2.5.1 does.

when is a path-based gesture *not* dragging, in this context? i'm failing to see the nuance here. also, regardless of that, if an author implements a path-based gesture in an attempt to provide an alternative, they'd still fail 2.5.1 then anyway, so not acceptable either way?

**Comment by guyhickling** [comment]
> when is a path-based gesture not dragging, in this context? i'm failing to see the nuance here.

A swipe gesture is a PBG, because it usually has to be done in a particular direction. With horizontal sliders, for instance, if the swipe is more vertically up or down than horizontal it may scroll the page instead of moving the slider. But a swipe is not a drag.

**Comment by patrickhlauke** [comment]
not sure i've ever seen a slider that you can also swipe instead of drag. and again, keeping in mind that if you did somehow do that, you'd still have to satisfy 2.5.1 which would then fail, no?

**Comment by guyhickling** [comment]
Ok, maybe sliders was a poor example. The point I was making is that PBGs are not the same as dragging. So there is is still an inherent contradiction in this SC that needs to be fixed. The fact that 2.5.1 also covers PBGs does not alter the fact that 2.5.7 contradicts itself.

Hence both need to be excluded, as in:

> .....can be operated by a single pointer without dragging or path based gestures....

**Comment by patrickhlauke** [comment]
given that 2.5.1 essentially forbids single pointer path-based gestures (unless they themselves provide a single pointer alternative that does not involve paths), does 2.5.7 also need to reiterate this? would this not be akin to constantly reminding authors in every pointer SC as well "but make sure there's a way to operate this with the keyboard too!" which we don't do because it's redundant?

**Comment by alastc** [comment]
> PBGs are not the same as dragging

Dragging is a superset of path-based gestures.

(Or is it easier to say PBGs are a sub-set of dragging?)

A gesture (path based or otherwise) involves dragging a pointer along a screen, so nothing needs to be excluded from the new SC.

**Comment by guyhickling** [comment]
> does 2.5.7 also need to reiterate this?

a) But why not include it? - we only save two words by omitting it, at the possible risk of loss of clarity. 

While we in this discussion are fully aware of the implications of 2.5.1, people in web dev who are not so familiar with the WCAG and all its nuances could easily fail to make the connection. They could read 2.5.7 and think that a PBG is an allowable way to solve their problem, forgetting (or not knowing) that 2.5.1 covers it. The WCAG is difficult enough for outsiders to understand (it has a reputation for that in the web dev industry) without making it even more complex just to save two words! 

b) Each SC should stand on its own independent of the others. Not showing the exclusion of PBGs makes 2.5.7 dependent on 2.5.1 to get the full picture of this scenario.

**Comment by patrickhlauke** [comment]
> > PBGs are not the same as dragging
> 
> Dragging is a superset of path-based gestures.
> 
> (Or is it easier to say PBGs are a sub-set of dragging?)
> 
> A gesture (path based or otherwise) involves dragging a pointer along a screen, so nothing needs to be excluded from the new SC.

dragging involves an element being dragged, while PBGs might not have an underlying element that's being dragged about (think for instance drawing an "L" shapes with your finger on a page to invoke some command. so sadly the categorisation is not as clean.

mulling this over some more, maybe a good example to think about in the context of @guyhickling's worry is a carousel that only allows dragging of a slide in order to advance to the next slide. authors then think they also need to implement a flick gesture that is a PBG, and missing that doing that would fail 2.5.1...

while i think the current wording is nice and clean "if you have dragging, make sure there's a single pointer method that doesn't involve dragging, unless dragging's essential", if we really think it'll add extra super duper clarity, then i guess it could include it - but it'll make the "unless" bit awkward/ambiguous?

> can be operated by a single pointer without dragging or relying on a path-based gesture, unless dragging is essential.

**Comment by mraccess77** [comment]
If SC 2.5.7 prohibits path-based gestures and 2.5.1 does as well then we'd have to fail both SC when they are used.   I think it makes sense to have one cover dragging and one cover PBG as some people may meet one and not the impact on users may actually be different and this might more clearly indicate what is actually supported or not.  I'd think we could add a note to the understanding telling people that if they are looking to solve dragging then they would also need to implement a non-PBG to pass 2.5.1.

**Comment by guyhickling** [comment]
Patrick is right about the "unless ... essential" clause - that would also need PBGs to be added, as in "unless dragging or a path-based gesture is essential."

Jonathan, I think you have the solution - do it using a green note on 2.5.7 referring back to PBGs being covered in 2.5.1. That would be sufficient to make sure developers and others are aware of the collective requirements of both SCs, which meets my concern. And it avoids disturbing the SC wording.

It would also allow us to do the reverse as well - place a green note on 2.5.1 referring to the dragging in 2.5.7. That will bring PBGs and dragging together, and present the whole unified picture in both SCs.

**Comment by alastc** [comment]
We discussed this today: https://www.w3.org/2021/02/09-ag-minutes.html#item04

There was push-back on changing the definition, with a key technical reason being that pointer-cancellation relies on the current meaning. (I think it's the Abort or Undo bullet, where you might need to perform a gesture/drag/movement to abort the function before completion.)

The complexity was acknowledged, so we are not rejecting the issue, but still looking for the best approach.

To try and covey the results of today's discussion I'll try to write this as conclusions prior to a group response:

<hr>

The definition of single pointer is not wrong, in that there are a variety of actions you can perform with a single pointer, including path-based-gestures.

2.5.1 Pointer Gestures intentionally uses that definition and then adds "without a path-based gesture" to apply the correct scope.

2.5.2 Pointer Cancellation also uses that definition in  the "Abort or Undo" bullet, where you might need to perform a gesture/drag/movement to abort the function before completion.

2.5.7 Dragging works in a similar way to 2.5.1, but qualifies as "by a single pointer without dragging".

So the question then might be: What is the overlap between pointer-gestures (not a defined term) and [dragging movement](https://w3c.github.io/wcag/guidelines/22/#dfn-dragging-movements)?

The [understanding doc for pointer gestures](https://w3c.github.io/wcag/understanding/pointer-gestures.html) spends quite a lot of time on what pointer gestures are, and they are intentionally different from a general dragging action. There is overlap, you could drag in the same way as a gesture, but the difference is that a gesture requires an intermediate point, dragging does not. Dragging requires something that follows the pointer, gestures do not.

Guy wrote:
> 2.5.7 expressly suggests PBGs as an alternative to dragging, which contradicts the intention of 2.5.1 which says they cannot be used without an alternative.

2.5.7 does not suggest PSBs as an alternative, and whether we agree on that or not, it would not be a problem if one interaction causes a fail of one SC but not the other.

If an interaction is a gesture, it will come under 2.5.1, if it meets the definition of dragging movement, it will come under 2.5.7. There might be some overlap, although I cannot think of an example at the moment.

> none of the three SCs that reference "single pointer" intend to allow PBGs as the alternative

Not an alternative, but 2.5.7 does rely on on the examples being part of single-pointer interactions.

Patrick wrote:
> my PR here #809 attempts retrospectively to square this circle, by providing a definition of "single pointer" and then separately expanding what gestures can be performed with a single pointer.

It was pointed out that it doesn't really change the meaning, and the examples still include PBGs. There was a worry about removing the "one point of contact", although I wasn't thinking quickly enough to point out the mouse aspect there. I think there is still a change that can be made, but it doesn't solve this issue (if it does need solving).

Guy wrote:
> do it using a note in the Understanding for 2.5.7 referring back to PBGs being covered in 2.5.1...  place a similar note in 2.5.1's Understanding referring to the dragging in 2.5.7.

We could do. Personally I don't see the need as most people get 'gestures' and 'dragging', it's only when you have odd or in-between cases that you need to dig into the examples.

There is already a section in the Intent of 2.5.7 which outlines the difference, is another block of text needed for this?

**Comment by patrickhlauke** [comment]
> The definition of single pointer is not wrong

the point of the clarification in #809 was to try and straighten out the fact that WCAG uses the term "pointer" (which is the input *device*, like the mouse, a finger on a touchscreen, a stylus, per https://www.w3.org/TR/pointerevents/#dfn-pointer) to mean the "input event" triggered *with* that pointer. a synecdoche. but perhaps that was too fine a point.

**Comment by alastc** [comment]
@patrickhlauke - I haven't given up on updated that definition yet, it just doesn't help with this issue (on our 2.2 critical path, as it were).

**Comment by DavidMacDonald** [comment]
@alastc 
I agree... It won't be too hard to add some disambiguation using what's already there, adding a heading and a couple of sentences (from your response)


Disambiguation of Dragging and Pointer Gestures

Dragging motions where the movement between the start and end points can go in any direction and is entirely non-prescriptive are not path-based.  The user could drag in the same way as a gesture, but the difference is that a gesture requires an intermediate point, dragging does not. Dragging requires something that follows the pointer, gestures do not. Dragging motions are covered in Success Criterion 2.5.7 Dragging.

Here's where I would put it.
image.png
![image](https://user-images.githubusercontent.com/2365411/107557553-c42e2100-6ba7-11eb-9e77-76182ab89863.png)

And then the same section in the Dragging understanding
image.png
![image](https://user-images.githubusercontent.com/2365411/107557571-cb552f00-6ba7-11eb-94cb-85507922d697.png)


I could do a branch and pull request.

**Comment by alastc** [comment]
Thanks @DavidMacDonald, it's in the survey.

**Comment by alastc** [comment]
Closing this issue as the understanding doc update was merged, but noting that @gradualclearing might take another pass at it.
